caffe2/operators/dropout_op_cudnn.cc

```diff
@@ -0,0 +1,234 @@
+    CUDNN_ENFORCE(cudnnCreateTensorDescriptor(&data_desc_));
+
+    CUDNN_ENFORCE(cudnnCreateDropoutDescriptor(&dropout_desc_));
+    CUDNN_ENFORCE(cudnnDropoutGetStatesSize(cudnn_wrapper_.inline_cudnn_handle(), &states_size_in_bytes_));
+  }
+
+  ~CuDNNDropoutOp() {
+    CUDNN_ENFORCE(cudnnDestroyTensorDescriptor(data_desc_));
+    CUDNN_ENFORCE(cudnnDestroyDropoutDescriptor(dropout_desc_));
+  }
+
+  template <typename T, typename M>
+  bool DoRunWithType();
+
+  bool RunOnDevice() override;
+
+ protected:
+  CuDNNWrapper cudnn_wrapper_;
+  cudnnTensorDescriptor_t data_desc_;
+  cudnnDropoutDescriptor_t dropout_desc_;
+
+  vector<TIndex> cudnn_input_dims_;
+
+  float ratio_;
+  bool is_test_;
+
+  size_t states_size_in_bytes_, reserve_space_size_in_bytes_;
+  // Input: X, Output: Y, mask, states
+};
+
+class CuDNNDropoutGradientOp final : public Operator<CUDAContext> {
+ public:
+  USE_OPERATOR_FUNCTIONS(CUDAContext);
+  CuDNNDropoutGradientOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<CUDAContext>(operator_def, ws),
+        cudnn_wrapper_(&context_),
+        ratio_(OperatorBase::GetSingleArgument<float>("ratio", 0.5)),
+        is_test_(OperatorBase::GetSingleArgument<int>("is_test", 0)) {
+    DCHECK_GE(ratio_, 0);
+    DCHECK_LT(ratio_, 1);
+    CUDNN_ENFORCE(cudnnCreateTensorDescriptor(&data_desc_));
+
+    CUDNN_ENFORCE(cudnnCreateDropoutDescriptor(&dropout_desc_));
+    CUDNN_ENFORCE(cudnnDropoutGetStatesSize(cudnn_wrapper_.inline_cudnn_handle(), &states_size_in_bytes_));
+  }
+
+  ~CuDNNDropoutGradientOp() {
+    CUDNN_ENFORCE(cudnnDestroyTensorDescriptor(data_desc_));
+  }
+
+  template <typename T, typename M>
+  bool DoRunWithType();
+
+  bool RunOnDevice() override;
+
+ protected:
+  CuDNNWrapper cudnn_wrapper_;
+  cudnnTensorDescriptor_t data_desc_;
+  cudnnDropoutDescriptor_t dropout_desc_;
+
+  vector<TIndex> cudnn_input_dims_;
+
+  float ratio_;
+  bool is_test_;
+
+  size_t states_size_in_bytes_, reserve_space_size_in_bytes_;
+  // Input: dY, states, Output: dX
+};
+
+template <typename T, typename M>
+bool CuDNNDropoutOp::DoRunWithType() {
+  const auto& X = Input(0);
+  auto* Y = Output(0);
+  auto* reserve = Output(1);
+
+  auto size_prod = 1;
+  for (auto dim : X.dims()) {
+    size_prod *= dim;
+  }
+
+  // Reshape tensor descriptors if necessary
+  if (X.dims() != cudnn_input_dims_) {
+    VLOG(1) << "Setting descriptors";
+    cudnn_input_dims_ = X.dims();
+    CUDNN_ENFORCE(cudnnSetTensor4dDescriptor(
+          data_desc_,
+          GetCudnnTensorFormat(StorageOrder::NCHW),
+          cudnnTypeWrapper<T>::type,
+          size_prod,
+          1,
+          1,
+          1));
+    // get the reserve space we need
+    CUDNN_ENFORCE(cudnnDropoutGetReserveSpaceSize(
+          data_desc_, &reserve_space_size_in_bytes_));
+    // store both reserve and states in the same tensor
+    vector<int> state_size{(reserve_space_size_in_bytes_ + states_size_in_bytes_ + sizeof(T)) / sizeof(T)};
```

Comment:
@slayton58 This is breaking windows build. (conversion from 'std::size_t' to 'int' requires a narrowing conversion) can you change the type of reserve_space_size_in_bytes_ & states_size_in_bytes_ to int ?